ES modules in service workers post (#5343)
GoogleChrome/web.dev: src/site/content/en/blog/es-modules-in-sw/index.md

modern library or configuration code that's shared with other runtimes that
support ES modules.

Attempting to share code in this way prior to ES modules entailed using older
"universal" module formats like [UMD](https://github.com/umdjs/umd) that include
unneeded boilerplate, and writing code that made changes to globally exposed
variables.

Scripts imported via ES modules can trigger the service worker
[update](https://developers.google.com/web/fundamentals/primers/service-workers/lifecycle#updates)
flow if their contents change, matching the
[behavior](https://developers.google.com/web/updates/2019/09/fresher-sw#checks_for_updates_to_imported_scripts)
of
<code>[importScripts()](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts)</code>.

## Current limitations

### Static imports only

ES modules can be imported in one of two ways: either
[statically](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import),
using the `import ... from '...'` syntax, or
[dynamically](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#dynamic_imports),

> using an `import('...')` statement. Inside of a service worker, only the static

Author: I've updated to switch to that language.

syntax is currently supported.

This limitation is analogous to a
[similar restriction](https://developers.google.com/web/updates/2018/10/tweaks-to-addAll-importScripts)
placed on `importScripts()` usage. Dynamic calls to `importScripts()` do not
work inside of a service worker, and all `importScripts()` calls, which are
inherently synchronous, must complete before the service worker completes its

> `install` phase. This restriction ensures that the browser knows about, and is

Author: `importScripts()` isn't supported in client pages. It's only exposed in the `WorkerGlobalScope`: https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts

able to implicitly cache, all JavaScript code needed for a service worker's
implementation during installation.

> At some point in the future, this restriction might be lifted, and dynamic ES

Reviewer: Omit needless words.

has been supported in Chrome and Edge since
[version 80](https://chromestatus.com/feature/5761300827209728), as well as
[recent versions](https://bugs.webkit.org/show_bug.cgi?id=164860) of Safari.
Both static and dynamic ES module imports are supported in dedicated workers.

Chrome and Edge have supported ES modules in
[shared workers](https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker)
since [version 83](https://chromestatus.com/feature/5169440012369920), but no
other browser offers support at this time.

### No support for import maps

[Import maps](https://github.com/WICG/import-maps/blob/main/README.md) allow
runtime environments to rewrite module specifiers, to, for example, prepend the
URL of a preferred CDN from which the ES modules can be loaded.

While Chrome and Edge
[version 89](https://www.chromestatus.com/feature/5315286962012160) and above
support import maps, they currently
[cannot be used](https://github.com/WICG/import-maps/issues/2) with service
workers.

## Browser support

ES modules in service workers are supported in Chrome and Edge starting with
[version 91](https://chromestatus.com/feature/4609574738853888).

Safari added support in the

> [Technology Preview 122 release](https://webkit.org/blog/11577/release-notes-for-safari-technology-preview-122/#:~:text=Added%20support%20for%20modules%20in%20Service%20Workers),

Reviewer: ```suggestion
[Technology Preview 122 Release](https://webkit.org/blog/11577/release-notes-for-safari-technology-preview-122/#:~:text=Added%20support%20for%20modules%20in%20Service%20Workers),
```

Reviewer: As a writer I think Apple is wrong in capitalizing 'release'. But they did and the rule is to follow their practice.

> and developers should expect to see this functionality released in the standard
> version of Safari in a future release.

Reviewer: ```suggestion
and developers should expect to see this functionality released in the stable
version of Safari in the future.
```

Firefox does not currently support this functionality, and updates on their
position can be found in this
[GitHub issue](https://github.com/mozilla/standards-positions/issues/499).

## Example code

This is a basic example of using a shared ES module in a web app's `window`
context, while also registering a service worker that uses the same ES module:

```javascript
// Inside config.js:
export const cacheName = 'my-cache';
```

```javascript
// Inside your web app:
<script type="module">
  import {cacheName} from './config.js';
  // Do something with cacheName.

  await navigator.serviceWorker.register('es-module-sw.js', {
    type: 'module',
  });
</script>
```

```javascript
// Inside es-module-sw.js:
import {cacheName} from './config.js';

self.addEventListener('install', (event) => {
  event.waitUntil((async () => {
    const cache = await caches.open(cacheName);
    // ...
  })());
});
```

### Backwards compatibility

The above example would work fine if all browsers supported ES modules in
service workers, but as of this writing, that's not the case.

To accommodate browsers that don't have built-in support, you can run your

> service worker script through a

Reviewer: ```suggestion
service worker script through an
```
